File: src/streambench/flinkbench/src/main/java/com/intel/flinkbench/microbench/DistinctCount.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.flinkbench.microbench;

import com.intel.flinkbench.datasource.StreamBase;
import com.intel.flinkbench.util.BenchLogUtil;
import com.intel.flinkbench.util.FlinkBenchConfig;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

import java.util.HashSet;
import java.util.Set;

public class DistinctCount extends StreamBase {
    Set<String> map = new HashSet<String>();
    @Override
    public void processStream(FlinkBenchConfig config) throws Exception {
        createDataStream(config);
        final String seperator = config.separator;
        DataStream<Tuple2<String, String>> dataStream = getDataStream();
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        dataStream
                .flatMap(new FlatMapFunction<Tuple2<String, String>, Tuple2<String, Tuple2<String, Integer>>>() {
                    @Override
                    public void flatMap(Tuple2<String, String> sentence, Collector<Tuple2<String, Tuple2<String, Integer>>> out) throws Exception {
                        for (String word : sentence.f1.split(seperator)) {
                            map.add(word);
                            out.collect(new Tuple2<String, Tuple2<String, Integer>>(sentence.f0, new Tuple2<String, Integer>(word, 1)));
                        }
                    }
                });
        BenchLogUtil.logMsg("stream distinct count: " + map.size());
        env.execute("distinct count job.");
    }
}

File: src/streambench/sparkbench/src/main/scala/com/intel/hibench/streambench/spark/application/DistinctCount.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.hibench.streambench.spark.application

import com.intel.hibench.streambench.common.{UserVisitParser, Logger}
import com.intel.hibench.streambench.spark.util.SparkBenchConfig

import org.apache.spark.streaming.dstream.DStream
import org.apache.spark.streaming.{StateSpec, State, StreamingContext}

class DistinctCount(config:SparkBenchConfig, logger: Logger)
  extends BenchRunnerBase(config, logger) {

  override def process(ssc: StreamingContext, lines: DStream[(Long, String)]) {
    val distinctFunc = (word: String, one: Option[Int], state: State[Int]) => {
      state.update(1)
      word
    }

    val parsedLine: DStream[(String, Int)] = lines.map(line => {
      val userVisit = UserVisitParser.parse(line._2)
      (userVisit.getIp, 1)
    })

    val distinctCount = parsedLine.mapWithState(StateSpec.function(distinctFunc)).stateSnapshots()
    if(config.debugMode) {
      distinctCount.print()
    } else {
      distinctCount.foreachRDD(rdd => rdd.foreach(_ => Unit))
    }
  }
}
